crates/wasmparser/src/readers/instance_section.rs

```diff
@@ -29,13 +29,14 @@ impl<'a> InstanceSectionReader<'a> {
             &self.reader.buffer[self.reader.position..],
             self.original_position(),
         )?;
-        self.reader.skip_bytes(1)?;
-        self.reader.skip_var_32()?;
-        let count = self.reader.read_var_u32()?;
-        for _ in 0..count {
-            self.reader.skip_bytes(1)?;
-            self.reader.skip_var_32()?;
+
+        // TODO: should probably figure out a different API for wasmparser such
+        // that we don't have to read instances to skip them here.
```

Comment:
My impression is that instance definitions are relatively small and fast to decode in comparison to the amount of computation necessary to perform an instantiation and thus a single thread straight-line decoding them should have no problem keeping many cores busy with instantiation work.  Also, I think the average number of instance defs will be low.  That being said, because instance sections can be repeated, an optimizing producer/consumer could always break up a big group of instance definitions into multiple sections ;)

Reply:
Ah yeah this is also more of a problem with wasmparser's APIs than the spec I think. The way we currently expose validation means that wasmtime, for example, parses the import section twice. First it hands the binary blob to the validator which iterates over everything and validates it. *Then* wasmtime iterates and parses it again to process it for wasmtime's own purposes.

This block with instantiation is similar to that where I think we just need to redesign some APIs in wasmparser (e.g. not relying on iterators so much in places like this). That's a larger scale change though and definitely of questionable benefit at this time because as Luke mentions it's all extremely fast to parse right now anyway

Reply:
https://github.com/bytecodealliance/wasm-tools/issues/188